physicsnemo/models/layers/ball_query.py

```diff
@@ -26,244 +237,91 @@ class BallQuery(torch.autograd.Function):
+        # there isn't checking that the devices match, but it will crash if they do not.
+        # try:
+        #     device = str(wp.get_device())
+        # except Exception:
+        #     device = "cuda"
 
-        try:
-            device = str(wp.get_device())
-        except Exception:
-            device = "cuda"
-
-        # Convert from torch to warp
-        ctx.points1 = wp.from_torch(
-            points1[0], dtype=wp.vec3, requires_grad=points1.requires_grad
-        )
-        ctx.points2 = wp.from_torch(
-            points2[0], dtype=wp.vec3, requires_grad=points2.requires_grad
-        )
-        # ctx.lengths1 = wp.from_torch(lengths1, dtype=wp.int32, requires_grad=False)
-        # ctx.lengths2 = wp.from_torch(lengths2, dtype=wp.int32, requires_grad=False)
         ctx.k = k
         ctx.radius = radius
 
-        # Allocate the mapping and outputs
-        mapping = torch.zeros(
-            1,
-            points1.shape[1],
-            k,
-            dtype=torch.int32,
-            device=device,
-            requires_grad=False,
-        )
-        ctx.mapping = wp.from_torch(mapping, dtype=wp.int32, requires_grad=False)
-        num_neighbors = torch.zeros(
-            1,
-            points1.shape[1],
-            dtype=torch.int32,
-            device=device,
-            requires_grad=False,
-        )
-        ctx.num_neighbors = wp.from_torch(
-            num_neighbors, dtype=wp.int32, requires_grad=False
-        )
-        outputs = torch.zeros(
-            1,
-            points1.shape[1],
-            k,
-            3,
-            dtype=torch.float32,
-            device=device,
-            requires_grad=(points1.requires_grad or points2.requires_grad),
-        )
-        ctx.outputs = wp.from_torch(
-            outputs,
-            dtype=wp.float32,
-            requires_grad=(points1.requires_grad or points2.requires_grad),
-        )
-
         # Make grid
         ctx.hash_grid = hash_grid
 
-        # Build the grid
-        ctx.hash_grid.build(ctx.points2, radius)
-
-        ctx.dim = [ctx.points1.shape[0]]
-
-        # Run the kernel to get mapping
-        wp.launch(
-            BallQuery.ball_query,
-            inputs=[
-                ctx.points1,
-                ctx.points2,
-                ctx.hash_grid.id,
-                k,
-                radius,
-            ],
-            outputs=[
-                ctx.mapping,
-                ctx.num_neighbors,
-            ],
-            dim=ctx.dim,
-        )
-
-        # Run the kernel to get outputs
-        wp.launch(
-            BallQuery.sparse_ball_query,
-            inputs=[
-                ctx.points2,
-                ctx.mapping,
-                ctx.num_neighbors,
-            ],
-            outputs=[
-                ctx.outputs,
-            ],
-            dim=ctx.dim,
+        # Apply the primitive.  Note the batch index is removed.
+        mapping, num_neighbors, outputs = _ball_query_forward_primative_(
```

Comment:
Super nit-pick, sorry
```suggestion
        mapping, num_neighbors, outputs = _ball_query_forward_primitive_(
```

Reply:
Yes, this one was pointed out to me to fix elsewhere too  :)